cli/waiter/util.py

```diff
@@ -12,6 +12,16 @@
 FALSE_STRINGS = ('no', 'false', 'n')
 
 
+def assoc_in(obj, keys, value):
+    """Given a list of keys and a value, return a new object with that value set"""
+    obj = obj.copy()
```

Comment:
It's not clear to me why you need to make a copy of the object here. If you don't really need the copy, maybe we just rename this `set_in` or `update_in` or something like that and not bother with the copy?

Otherwise, you need to copy at every level if you're trying to avoid mutating the original (nested) map:
```
>>> def assoc_in(obj, keys, value):
...     """Given a list of keys and a value, return a new object with that value set"""
...     obj = obj.copy()
...     cur_node = obj
...     for key in keys[:-1]:
...         cur_node = cur_node.setdefault(key, {})
...     cur_node[keys[-1]] = value
...     return obj
...
>>> assoc_in(m, ['e'], 30)
{'a': {'b': {'c': 1}, 'd': 20}, 'e': 30}
>>> m
{'a': {'b': {'c': 1}, 'd': 20}, 'e': 3}
>>> m = dict(a=dict(b=dict(c=1), d=2), e=3)
>>> m
{'a': {'b': {'c': 1}, 'd': 2}, 'e': 3}
>>> assoc_in(m, ['a', 'd'], 20)
{'a': {'b': {'c': 1}, 'd': 20}, 'e': 3}
>>> m
{'a': {'b': {'c': 1}, 'd': 20}, 'e': 3}
```

Reply:
good point, I didn't realize that .copy() doesn't make a deep copy. Going to make this change